### PR TITLE
New option style

### PR DESCRIPTION
**Pull request by kuenishi**
Code will follow discussion,  or time.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,35 +33,86 @@ Ham = msgpack:pack(Spam),
...
+```
+mode        as_binary    as_list
+-----------+------------+-------
+packing
+binary()    str          bin
+string()    bin*         str*
+list()      array        array
+-----------+------------+-------
+unpacking
+bin         string()     binary()
```
Comment (mururu) [inline README.md]: I think `bin` type should be unpacked as `binary()` when `as_binary` option is set.

**Comment by mururu** [comment]
It is better to refer to the default value of each options.

File: README.md (hunk 1 above)
Comment (kuenishi) [inline README.md]: I should have written note here. With this decode case I intended trying to decode with UTF-8 validation, and if it is invalid as a string it falls back to `binary()`. I still try to keep symmetry.

Comment (mururu) [inline README.md]: The behavior of unpacking for bin type should not change based on whether or not it is valid as a representation of UTF-8. bin type is just a byte array.
I understand that you want to keep it symmetry, but in this case it is not reasonable to me.

Comment (kuenishi) [inline README.md]: I understand that. What about packing `string()` with `as_binary` ? This case involves UTF-8 validation to know whether a list is a string or not.

**Comment by kuenishi** [comment]
Did some update.

**Comment by mururu** [comment]
Ah, fair enough to me. Great work!

**Comment by kuenishi** [comment]
@mururu Thank you for review. It helped me brush up the spec a lot.

Will find time to implement this - but other review comments would still be welcomed.

**Comment by kuenishi** [comment]
TODO: add tests on new bin/str/spec combinations.

**Pull request merged**